### PR TITLE
Adding request logging handler.

### PR DESCRIPTION
**Pull request by taras-kobernyk-localstack**
This logger is supposed to help with 2 things:
1. It is easier to debug configurations when we see what they are. If LocalStack is queried by Terraform etc, it is not always clear, what calls are there, what resources get created. This logging would make it explicit.
2. Not all calls end up logging regular responses. For example, test_crypto.py from the Pro version uses some streaming libraries that make calls that do not result in ResponseLogger outputting anything. This RequestLogger succeeds there.

The resulting log looks like this:

AWS request to kms.DescribeKey. Can reproduce in CLI with: awslocal kms describe-key --region us-east-1 --cli-input-json '{"KeyId": "41cc5e04-7326-43c6-b7c5-79cfedb23f01"}'
AWS kms.DescribeKey => 200

There is an assumption here that all operation names like GenerateDataKey can be turned into proper operation names for CLI with this parsing - splitting CamelCase, lowercasing all the words and joining them with dashes. But even if that is not true - there is already a benefit from seeing parameters.

### CODE REVIEW

**Review by thrau — COMMENTED** [review]
i think there are some conceptual things to clarify before we continue working on this.

i can make out two key ideas in this PRs:
* logging requests as they arrive, rather than as part of the response logger
* additional output for helping developers understand requests

in principle i find both of those ideas good.


however there are multiple problems with the PR in its current state.

* the logging becomes very verbose. we're adding `Can reproduce in CLI with: awslocal ...` for *every* request, which seems a bit excessive to me tbh. i very rarely care about this, and can see that this will be particularly annoying in CI logs.
* we're logging requests twice now. once as it comes in, and then as it goes through the response logger. i don't think that's helpful
* we're calling `restore_payload` for *every* request, which is a relatively expensive operation and should only be called when really necessary
* this doesn't seem to work for services that use the query protocol. for example:
`awslocal sqs list-queues` translates to 
`AWS request to sqs.ListQueues. Can reproduce in CLI with: awslocal sqs list-queues --region us-east-1 --cli-input-json 'Action=ListQueues&Version=2012-11-05'`, and that CLI command will obviously not work since it's not json.
* there seems to be an issue when the trace logger is enabled (`LS_LOG=trace`), where the log is full of these exceptions:
  ```
  2022-09-24T20:04:08.090 DEBUG --- [   asgi_gw_0] l.aws.serving.wsgi         : POST localhost:4566/
  --- Logging error ---
  Traceback (most recent call last):
    File "/home/thomas/.pyenv/versions/3.10.4/lib/python3.10/logging/__init__.py", line 1100, in emit
      msg = self.format(record)
    File "/home/thomas/.pyenv/versions/3.10.4/lib/python3.10/logging/__init__.py", line 943, in format
      return fmt.format(record)
    File "/home/thomas/workspace/localstack/localstack/localstack/logging/format.py", line 140, in format
      record.input = self._copy_service_dict(record.input)
  AttributeError: 'LogRecord' object has no attribute 'input'
  Call stack:
    File "/opt/JetBrains/Toolbox/apps/PyCharm-P/ch-0/222.3739.56/plugins/python/helpers/pydev/_pydev_bundle/pydev_monkey.py", line 798, in __call__
      ret = self.original_func(*self.args, **self.kwargs)
    File "/home/thomas/.pyenv/versions/3.10.4/lib/python3.10/threading.py", line 966, in _bootstrap
      self._bootstrap_inner()
    File "/home/thomas/.pyenv/versions/3.10.4/lib/python3.10/threading.py", line 1009, in _bootstrap_inner
      self.run()
    File "/home/thomas/.pyenv/versions/3.10.4/lib/python3.10/threading.py", line 946, in run
      self._target(*self._args, **self._kwargs)
    File "/home/thomas/.pyenv/versions/3.10.4/lib/python3.10/concurrent/futures/thread.py", line 83, in _worker
      work_item.run()
    File "/home/thomas/.pyenv/versions/3.10.4/lib/python3.10/concurrent/futures/thread.py", line 58, in run
      result = self.fn(*self.args, **self.kwargs)
    File "/home/thomas/workspace/localstack/localstack/localstack/aws/serving/wsgi.py", line 50, in __call__
      self.gateway.process(request, response)
    File "/home/thomas/workspace/localstack/localstack/localstack/aws/gateway.py", line 36, in process
      chain.handle(context, response)
    File "/home/thomas/workspace/localstack/localstack/localstack/aws/chain.py", line 90, in handle
      handler(self, self.context, response)
    File "/home/thomas/workspace/localstack/localstack/localstack/aws/handlers/logging.py", line 158, in __call__
      self._log(context, response)
    File "/home/thomas/workspace/localstack/localstack/localstack/aws/handlers/logging.py", line 183, in _log
      aws_logger.info(
  Message: "AWS request to %s.%s. Can reproduce in CLI with: awslocal %s %s --region %s --cli-input-json '%s'"
  Arguments: ('sqs', 'ListQueues', 'sqs', 'list-queues', 'us-east-1', 'Action=ListQueues&Version=2012-11-05')
  --- Logging error ---
  Traceback (most recent call last):
    File "/home/thomas/.pyenv/versions/3.10.4/lib/python3.10/logging/__init__.py", line 1100, in emit
      msg = self.format(record)
    File "/home/thomas/.pyenv/versions/3.10.4/lib/python3.10/logging/__init__.py", line 943, in format
      return fmt.format(record)
    File "/home/thomas/workspace/localstack/localstack/localstack/logging/format.py", line 140, in format
      record.input = self._copy_service_dict(record.input)
  AttributeError: 'LogRecord' object has no attribute 'input'
  Call stack:
    File "/opt/JetBrains/Toolbox/apps/PyCharm-P/ch-0/222.3739.56/plugins/python/helpers/pydev/_pydev_bundle/pydev_monkey.py", line 798, in __call__
      ret = self.original_func(*self.args, **self.kwargs)
    File "/home/thomas/.pyenv/versions/3.10.4/lib/python3.10/threading.py", line 966, in _bootstrap
      self._bootstrap_inner()
    File "/home/thomas/.pyenv/versions/3.10.4/lib/python3.10/threading.py", line 1009, in _bootstrap_inner
      self.run()
    File "/home/thomas/.pyenv/versions/3.10.4/lib/python3.10/threading.py", line 946, in run
      self._target(*self._args, **self._kwargs)
    File "/home/thomas/.pyenv/versions/3.10.4/lib/python3.10/concurrent/futures/thread.py", line 83, in _worker
      work_item.run()
    File "/home/thomas/.pyenv/versions/3.10.4/lib/python3.10/concurrent/futures/thread.py", line 58, in run
      result = self.fn(*self.args, **self.kwargs)
    File "/home/thomas/workspace/localstack/localstack/localstack/aws/serving/wsgi.py", line 50, in __call__
      self.gateway.process(request, response)
    File "/home/thomas/workspace/localstack/localstack/localstack/aws/gateway.py", line 36, in process
      chain.handle(context, response)
    File "/home/thomas/workspace/localstack/localstack/localstack/aws/chain.py", line 90, in handle
      handler(self, self.context, response)
    File "/home/thomas/workspace/localstack/localstack/localstack/aws/handlers/logging.py", line 158, in __call__
      self._log(context, response)
    File "/home/thomas/workspace/localstack/localstack/localstack/aws/handlers/logging.py", line 183, in _log
      aws_logger.info(
  Message: "AWS request to %s.%s. Can reproduce in CLI with: awslocal %s %s --region %s --cli-input-json '%s'"
  Arguments: ('sqs', 'ListQueues', 'sqs', 'list-queues', 'us-east-1', 'Action=ListQueues&Version=2012-11-05')
  ```

key idea 1: if we want to split request/repsonse logging in a generic way), i think we need to spend a bit more time thinking about what the log output should look like. logging a request twice is too verbose IMO. if we separate request and response logging, then we need to a) completely rework the log formatters in `format.py`, and b) introduce a way to associate requests to responses in the logs (with request IDs), and add those to the logs. otherwise it won't be possible to know which request belongs to which response.  @dfangl and i spent quite a bit of time thinking about logging for the v1 release, and found that to not be a good way forward. as i said, i'm not against it per-se, if you can come up with a good solution. however this seems to be a bit out of scope of this PR, since it will be quite a significant rework.

key idea 2: i would be OK merging this if we keep this logger behind an opt-in feature flag, something like `LOG_AWS_REQUEST_CLI_REPRODUCERS=1` (or something specific along those lines), and the problems with `LS_LOG=trace` and the other AWS protocols are fixed.

in summary, i would try to make the PR more specific to key idea 2. by first fixing all the issues, and then move away from the idea to introduce a second generic request logging approach.

**Comment by taras-kobernyk-localstack** [comment]
I wanted this to only happen for DEBUG level logging. My bad that I used INFO. Using DEBUG should take care of some of these concerns - like logging being too verbose or too expensive.

If by logging requests twice you mean that in response logging we log something like
    AWS kms.DescribeKey => 200
then I am trying to do something else. Simple mention of a request happening is not enough in some case - like those I have mentioned in the PR description. Need full settings somewhere as well.

Matching request logging to response logging would have been useful, sure. I was actually hoping that there was some easy mechanism for you to recommend me. :-) But we can probably just generate some UUID and include it into our context. And then use it in both places.

The current logging of mine might be too verbose, true. I was just trying to imagine the use cases for this logging. And at least for me it was useful to be able to reproduce the call by myself - using CLI, for example. So I was formatting the logs the corresponding way - so I was able to simply copy&paste the output. And the `Can reproduce in CLI with` part was just an attempt to make this use case obvious to whoever is going to bump into these logs unprepared. Can try to rephrase and make it shorter, sure.

The issue with logs like `'Action=ListQueues&Version=2012-11-05` is real. Unfortunately, I only saw such logs after creating this PR. Also there are cases where `restore_payload ` returns nothing. So this PR requires more work to take those cases into account.
